client/src/main/java/com/netflix/conductor/client/http/PayloadStorage.java

```diff
@@ -0,0 +1,148 @@
+ * limitations under the License.
+ */
+
+package com.netflix.conductor.client.http;
+
+import com.amazonaws.util.IOUtils;
+import com.netflix.conductor.client.exceptions.ConductorClientException;
+import com.netflix.conductor.common.run.ExternalStorageLocation;
+import com.netflix.conductor.common.utils.ExternalPayloadStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+
+/**
+ * An implementation of {@link ExternalPayloadStorage} for storing large JSON payload data.
+ */
+class PayloadStorage implements ExternalPayloadStorage {
+    private static final Logger logger = LoggerFactory.getLogger(PayloadStorage.class);
+
+    private final ClientBase clientBase;
+
+    PayloadStorage(ClientBase clientBase) {
+        this.clientBase = clientBase;
+    }
+
+    /**
+     * This method is not intended to be used in the client.
+     * The client makes a request to the server to get the {@link ExternalStorageLocation}
+     */
+    @Override
+    public ExternalStorageLocation getLocation(Operation operation, PayloadType payloadType, String path) {
+        String uri;
+        switch (payloadType) {
+            case WORKFLOW_INPUT:
+            case WORKFLOW_OUTPUT:
+                uri = "workflow";
+                break;
+            case TASK_INPUT:
+            case TASK_OUTPUT:
+                uri = "tasks";
+                break;
+            default:
+                throw new ConductorClientException(String.format("Invalid payload type: %s for operation: %s", payloadType.toString(), operation.toString()));
+        }
+        return clientBase.getForEntity(String.format("%s/externalstoragelocation", uri), new Object[]{"path", path}, ExternalStorageLocation.class);
+    }
+
+    /**
+     * Uploads the payload to the uri specified.
+     *
+     * @param uri         the location to which the object is to be uploaded
+     * @param payload     an {@link InputStream} containing the json payload which is to be uploaded
+     * @param payloadSize the size of the json payload in bytes
+     * @throws ConductorClientException if the upload fails due to an invalid path or an error from external storage
+     */
+    @Override
+    public void upload(String uri, InputStream payload, long payloadSize) {
+        HttpURLConnection connection = null;
+        try {
+            URL url = new URI(uri).toURL();
+
+            connection = (HttpURLConnection) url.openConnection();
+            connection.setDoOutput(true);
+            connection.setRequestMethod("PUT");
+
+            try (BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(connection.getOutputStream())) {
+                IOUtils.copy(payload, bufferedOutputStream);
+
```

Comment:
What happens if there is a connection timeout? You might want to retry?

Reply:
What chunk size does the stream uses internally?

Reply:
Why not just you S3 APIs to upload and download?

Reply:
IOUtils handles the buffering of the stream.
This is the implementation within the client which is unaware of the ExternalPayloadStorage that is used, this is decided by the server.